File: app/javascript/channels/borrower_message_channel.js
import consumer from "channels/consumer"

document.addEventListener('turbo:load', () => {
  const messageContainer = document.getElementById('message-list')
  if (!messageContainer) return

  const applicationId = messageContainer.dataset.applicationId
  if (!applicationId) return

  const subscription = consumer.subscriptions.create(
    { channel: 'BorrowerMessageChannel', application_id: applicationId },
    {
      connected() {
        console.log('✓ Connected to BorrowerMessageChannel')
      },

      disconnected() {
        console.log('✗ Disconnected from BorrowerMessageChannel')
      },

      received(data) {
        // Add new message to the list
        const messageEl = createMessageElement(data)
        messageContainer.appendChild(messageEl)
        
        // Scroll to bottom
        messageContainer.scrollTop = messageContainer.scrollHeight
        
        // Flash badge on new message (if from other party)
        if (!data.is_current_user) {
          flashNewMessageBadge()
        }
      },

      send_message(message) {
        this.perform('send_message', { message: message })
      }
    }
  )

  // Handle message form submission
  const messageForm = document.getElementById('message-form')
  if (messageForm) {
    messageForm.addEventListener('submit', (e) => {
      e.preventDefault()
      const textarea = messageForm.querySelector('textarea[name="message"]')
      const message = textarea.value.trim()
      
      if (message.length > 0) {
        subscription.send_message(message)
        textarea.value = ''
        textarea.focus()
        updateCharCount(textarea)
      }
    })

    // Character counter
    const textarea = messageForm.querySelector('textarea[name="message"]')
    if (textarea) {
      textarea.addEventListener('input', () => updateCharCount(textarea))
    }
  }

  // Mark messages as read when viewing
  markMessagesAsRead()
})

function createMessageElement(data) {
  const wrapper = document.createElement('div')
  wrapper.className = `message-item message-from-${data.sender_type}`
  wrapper.innerHTML = `
    <div class="message-avatar">
      <img src="${data.user_avatar}" alt="${data.user_name}" class="avatar-img">
    </div>
    <div class="message-content">
      <div class="message-header">
        <span class="message-author">${escapeHtml(data.user_name)}</span>
        <span class="message-time">${data.created_at}</span>
      </div>
      <div class="message-text">${escapeHtml(data.message)}</div>
    </div>
  `
  return wrapper
}

function updateCharCount(textarea) {
  const count = textarea.value.length
  const limit = 5000
  const counter = textarea.parentElement.querySelector('.char-count')
  if (counter) {
    counter.textContent = `${count}/${limit}`
    counter.classList.toggle('warning', count > limit * 0.9)
  }
}

function markMessagesAsRead() {
  // Find all unread messages from lender and mark as read
  document.querySelectorAll('[data-message-id][data-unread="true"]').forEach(el => {
    const messageId = el.dataset.messageId
    fetch(`/borrower/messages/${messageId}`, {
      method: 'PATCH',
      headers: {
        'Content-Type': 'application/json',
        'X-CSRF-Token': document.querySelector('meta[name="csrf-token"]').content
      },
      body: JSON.stringify({ mark_as_read: true })
    })
    el.dataset.unread = 'false'
  })
}

function flashNewMessageBadge() {
  // Optional: add visual notification for new message
  const badge = document.querySelector('[data-unread-count]')
  if (badge) {
    badge.classList.add('pulse')
    setTimeout(() => badge.classList.remove('pulse'), 1000)
  }
}

function escapeHtml(text) {
  const div = document.createElement('div')
  div.textContent = text
  return div.innerHTML
}

File: app/channels/borrower_message_channel.rb
class BorrowerMessageChannel < ApplicationCable::Channel
  def subscribed
    @application = Application.find(params[:application_id])
    authorize_access!

    stream_for @application

    # Mark all messages from other party as read when user subscribes
    @application.borrower_messages.unread.where.not(user_id: current_user.id).each(&:mark_as_read!)
  end

  def unsubscribed
    # Any cleanup when user leaves the channel
  end

  def send_message(data)
    @application = Application.find(params[:application_id])
    authorize_access!

    # Determine sender type (borrower or lender)
    sender_type = @application.user_id == current_user.id ? :borrower : :lender
    lender = sender_type == :lender ? current_user : nil

    message = @application.borrower_messages.create!(
      user_id: current_user.id,
      lender_id: lender&.id,
      message: data["message"],
      sender_type: sender_type
    )

    # Broadcast to all subscribers of this application's channel
    BorrowerMessageChannel.broadcast_to(
      @application,
      {
        id: message.id,
        user_name: current_user.full_name,
        user_avatar: avatar_url(current_user),
        sender_type: sender_type,
        message: message.message,
        created_at: message.created_at.strftime("%H:%M %p"),
        is_current_user: true
      }
    )
  end

  private

  def authorize_access!
    # Only borrower and assigned lender can access this channel
    is_borrower = @application.user_id == current_user.id
    is_lender = @application.lender_id == current_user.id

    reject unless is_borrower || is_lender
  end

  def avatar_url(user)
    # Return initials-based avatar or gravatar
    "https://api.dicebear.com/7.x/initials/svg?seed=#{user.full_name}"
  end
end
